src/ConfigSpace/hyperparameters/hyperparameter.py

```diff
@@ -0,0 +1,780 @@
+from dataclasses import dataclass, field, replace
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Generic,
+    Literal,
+    TypeVar,
+    overload,
+)
+from typing_extensions import Self, deprecated
+
+import numpy as np
+
+from ConfigSpace.types import DType, Number, ValueT, f64, i64
+
+if TYPE_CHECKING:
+    from ConfigSpace.hyperparameters.distributions import Distribution
+    from ConfigSpace.hyperparameters.hp_components import Neighborhood, Transformer
+    from ConfigSpace.hyperparameters.uniform_float import UniformFloatHyperparameter
+    from ConfigSpace.hyperparameters.uniform_integer import UniformIntegerHyperparameter
+    from ConfigSpace.types import Array, Mask
+
+
+@dataclass(init=False)
+class Hyperparameter(ABC, Generic[ValueT, DType]):
+    """Base class for all hyperparameters in the configuration space.
+
+    Please see the [reference page](../../../reference/hyperparameters.md) for more.
+    """
+
+    ORDERABLE: ClassVar[bool] = False
+    """If the hyperparameter values have an order. This is used for
+    conditionals and forbiddens relying on relationships.
+    """
+
+    LEGAL_VALUE_TYPES: ClassVar[tuple[type, ...] | Literal["all"]] = "all"
+    """The types of values that are legal for this hyperparameter. If set to
+    `"all"` any type is legal. Otherwise, a tuple of types can be provided.
+    """
+
+    name: str
+    """Name of the hyperparameter, with which it can be accessed."""
+
+    default_value: ValueT
+    """The default value of this hyperparameter."""
+
+    meta: Mapping[Hashable, Any] | None
+    """Field for holding meta data provided by the user. Not used by the ConfigSpace."""
+
+    size: int | float
+    """Size of the hyperparameter. For integer and choice hyperparameters this
+    is the number of possible values the hyperparameter can take on within the
+    specified range. For continuous hyperparameters this is usually `np.inf`.
+    """
+
+    _vector_dist: Distribution = field(repr=False)
+    _normalized_default_value: f64 = field(repr=False)
+    _transformer: Transformer[DType] = field(repr=False)
+    _neighborhood: Neighborhood = field(repr=False, compare=False)
+    _value_cast: Callable[[DType], ValueT] | None = field(repr=False, compare=False)
+    _neighborhood_size: float | Callable[[ValueT | DType | None], int | float] = field(
+        repr=False,
+        compare=False,
+    )
+
+    def __init__(
+        self,
+        name: str,
+        default_value: ValueT,
+        vector_dist: Distribution,
+        transformer: Transformer[DType],
+        neighborhood: Neighborhood,
+        size: int | float,
+        neighborhood_size: float | int | Callable[[DType | ValueT | None], int | float],
+        value_cast: Callable[[DType], ValueT] | None,
+        meta: Mapping[Hashable, Any] | None = None,
+    ) -> None:
+        """Initialize a hyperparameter.
+
+        Args:
+            name:
```

Comment:
Is it necessary to duplicate this information here?

Reply:
Not entirely but I don't have it a different way. This is documenting parameters, while the other ones are documenting attributes. I would say it's necessary as it shows up in editors as help information